File: c05_4_pruning_the_search.py
import unittest

dv=((1,0),(-1,0),(0,1),(0,-1))

def can_move(y,x,n,visited):
    return 0<=y<n and 0<=x<n and not visited[y][x]

def basic(n):
    count=0
    visited=[[False]*n for _ in range(n)]
    def rec(k,y,x):
        nonlocal count
        if not can_move(y,x,n,visited):
            return
        if k>=n*n and y==n-1 and x==n-1:
            count+=1
        visited[y][x]=True
        for dy,dx in dv:
            rec(k+1,y+dy,x+dx)
        visited[y][x]=False
    rec(1,0,0)
    return count

def optimize1(n):
    if n<2:
        return max(n,0)
    count=0
    visited=[[False]*n for _ in range(n)]
    visited[0][0]=True
    def rec(k,y,x):
        nonlocal count
        if not can_move(y,x,n,visited):
            return
        if k>=n*n and y==n-1 and x==n-1:
            count+=1
        visited[y][x]=True
        for dy,dx in dv:
            rec(k+1,y+dy,x+dx)
        visited[y][x]=False
    rec(2,1,0)
    return count*2

def optimize2(n):
    if n<2:
        return max(n,0)
    count=0
    visited=[[False]*n for _ in range(n)]
    visited[0][0]=True
    def rec(k,y,x):
        nonlocal count
        if not can_move(y,x,n,visited):
            return
        if y==n-1 and x==n-1:
            if k>=n*n:
                count+=1
            return
        visited[y][x]=True
        for dy,dx in dv:
            rec(k+1,y+dy,x+dx)
        visited[y][x]=False
    rec(2,1,0)
    return count*2

def optimize3(n):
    if n<2:
        return max(n,0)
    count=0
    visited=[[False]*n for _ in range(n)]
    visited[0][0]=True
    def rec(k,y,x):
        nonlocal count
        if not can_move(y,x,n,visited):
            return
        if y==n-1 and x==n-1:
            if k>=n*n:
                count+=1
            return
        if y in (0,n-1) and can_move(y,x-1,n,visited) and can_move(y,x+1,n,visited):
            return
        if x in (0,n-1) and can_move(y-1,x,n,visited) and can_move(y+1,x,n,visited):
            return
        visited[y][x]=True
        for dy,dx in dv:
            rec(k+1,y+dy,x+dx)
        visited[y][x]=False
    rec(2,1,0)
    return count*2

def optimize4(n):
    if n<2:
        return max(n,0)
    count=0
    visited=[[False]*n for _ in range(n)]
    visited[0][0]=True
    def rec(k,y,x):
        nonlocal count
        if not can_move(y,x,n,visited):
            return
        if y==n-1 and x==n-1:
            if k>=n*n:
                count+=1
            return
        if can_move(y+1,x,n,visited)==can_move(y-1,x,n,visited) and can_move(y,x+1,n,visited)==can_move(y,x-1,n,visited):
            return
        visited[y][x]=True
        for dy,dx in dv:
            rec(k+1,y+dy,x+dx)
        visited[y][x]=False
    rec(2,1,0)
    return count*2

class Test(unittest.TestCase):
    def test(self):
        for o in (basic,optimize1,optimize2,optimize3,optimize4):
            with self.subTest(o=o):
                self.assertEqual(o(1),1)
                self.assertEqual(o(5),104)

if __name__=='__main__':
    unittest.main()

File: c++/c05_4_pruning_the_search.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>
using namespace std;
using ll = long long;

namespace {
	const vector<pair<ll, ll>> dv = { {1, 0}, {-1, 0}, {0, 1}, {0, -1} };

	bool can_move(ll y, ll x, ll n, const vector<vector<bool>>& visited) {
		return 0 <= y && y < n && 0 <= x && x < n && !visited[y][x];
	}

	ll basic(ll n) {
		ll count = 0;
		vector<vector<bool>> visited(n, vector<bool>(n));
		auto rec = [&](this auto self, ll k, ll y, ll x) {
			if (!can_move(y, x, n, visited))
				return;
			if (k >= n * n && y == n - 1 && x == n - 1)
				++count;
			visited[y][x] = true;
			for (auto [dy, dx] : dv)
				self(k + 1, y + dy, x + dx);
			visited[y][x] = false;
			};
		rec(1, 0, 0);
		return count;
	}

	ll optimize1(ll n) {
		if (n < 2)
			return max(n, 0ll);
		ll count = 0;
		vector<vector<bool>> visited(n, vector<bool>(n));
		visited[0][0] = true;
		auto rec = [&](this auto self, ll k, ll y, ll x) {
			if (!can_move(y, x, n, visited))
				return;
			if (k >= n * n && y == n - 1 && x == n - 1)
				++count;
			visited[y][x] = true;
			for (auto [dy, dx] : dv)
				self(k + 1, y + dy, x + dx);
			visited[y][x] = false;
			};
		rec(2, 1, 0);
		return count * 2;
	}

	ll optimize2(ll n) {
		if (n < 2)
			return max(n, 0ll);
		ll count = 0;
		vector<vector<bool>> visited(n, vector<bool>(n));
		visited[0][0] = true;
		auto rec = [&](this auto self, ll k, ll y, ll x) {
			if (!can_move(y, x, n, visited))
				return;
			if (y == n - 1 && x == n - 1) {
				if (k >= n * n)
					++count;
				return;
			}
			visited[y][x] = true;
			for (auto [dy, dx] : dv)
				self(k + 1, y + dy, x + dx);
			visited[y][x] = false;
			};
		rec(2, 1, 0);
		return count * 2;
	}

	ll optimize3(ll n) {
		if (n < 2)
			return max(n, 0ll);
		ll count = 0;
		vector<vector<bool>> visited(n, vector<bool>(n));
		visited[0][0] = true;
		auto rec = [&](this auto self, ll k, ll y, ll x) {
			if (!can_move(y, x, n, visited))
				return;
			if (y == n - 1 && x == n - 1) {
				if (k >= n * n)
					++count;
				return;
			}
			if ((y == 0 || y == n - 1) && can_move(y, x - 1, n, visited) && can_move(y, x + 1, n, visited))
				return;
			if ((x == 0 || x == n - 1) && can_move(y - 1, x, n, visited) && can_move(y + 1, x, n, visited))
				return;
			visited[y][x] = true;
			for (auto [dy, dx] : dv)
				self(k + 1, y + dy, x + dx);
			visited[y][x] = false;
			};
		rec(2, 1, 0);
		return count * 2;
	}

	ll optimize4(ll n) {
		if (n < 2)
			return max(n, 0ll);
		ll count = 0;
		vector<vector<bool>> visited(n, vector<bool>(n));
		visited[0][0] = true;
		auto rec = [&](this auto self, ll k, ll y, ll x) {
			if (!can_move(y, x, n, visited))
				return;
			if (y == n - 1 && x == n - 1) {
				if (k >= n * n)
					++count;
				return;
			}
			if (can_move(y + 1, x, n, visited) == can_move(y - 1, x, n, visited) && can_move(y, x + 1, n, visited) == can_move(y, x - 1, n, visited))
				return;
			visited[y][x] = true;
			for (auto [dy, dx] : dv)
				self(k + 1, y + dy, x + dx);
			visited[y][x] = false;
			};
		rec(2, 1, 0);
		return count * 2;
	}

	void test(auto f) {
		EXPECT_EQ(f(1), 1);
		EXPECT_EQ(f(5), 104);
	}
}

TEST(C054PruningTheSearch, basic) {
	test(basic);
}

TEST(C054PruningTheSearch, optimize1) {
	test(optimize1);
}

TEST(C054PruningTheSearch, optimize2) {
	test(optimize2);
}

TEST(C054PruningTheSearch, optimize3) {
	test(optimize3);
}

TEST(C054PruningTheSearch, optimize4) {
	test(optimize4);
}
